### ISSUE TITLE
blueprint discussion about materials and per-material fields

### ISSUE THREAD

**Issue by xjrc** [LLNL/conduit#102]
This issue will house discussions related to the specification of materials and per-material fields in Blueprint.  For information about preliminary support for this data in Blueprint, see issue #93.

**Comment by xjrc** [comment]
Here's one approach that I came up with for materials and per-material fields:

### Materials as Fields ###

The basic idea behind this approach is to treat each material as a field and to allow fields to be defined over other fields.  Here's an example mesh using this scheme:

```
/* coordsets, topologies */
fields: {
  material0: {
    type: "material",  // special type for materials; effectively alternative to "scalar"
    association: "element",
    topology: "topology0",
    values: { /* volume fractions for material 0 */ },
  },
  /* more materials here */
  field0: {
    type: "scalar",
    association: "element",
    topology: "topology0",
    values: { /* uniform field values for all materials on "topology0" */ }
  },
  field1: {
    type: "scalar",
    association: "field",  // indicates that this field is defined over other fields
    topology: "topology0",  // perhaps unnecessary for a sub-field
    values: {
      material0: { /* field values defined over material 0 */ },
      /* ... */
      materialN: { /* field values defined over material N */ },
      /* all missing materials from this mcarray are assumed to have 0 values for this field */
    }
  }
}
```

I like this approach because I think it classifies materials properly (material volume fractions are just a type of field) and because it can be easily extended to support per-material fields and material species.  Unfortunately, this schema makes it difficult to define an mcarray of materials, though this could be fixed with some minor adjustments.

**Comment by xjrc** [comment]
Fixed with the upcoming changes in #185.
